napari/layers/labels/labels.py

```diff
@@ -1475,6 +1451,34 @@ def data_setitem(self, indices, value, refresh=True):
         # update the labels image
         self.data[indices] = value
 
+        if not (
+            isinstance(self.data, np.ndarray)  # numpy array, or
+            or hasattr(self.data, 'data')
+            and type(self.data.data) is np.ndarray  # numpy-backed xarray
+        ):
+            # In the absence of slicing, the current slice becomes
+            # invalidated by data_setitem; only in the special case of a NumPy
+            # array, or a NumPy-array-backed Xarray, is the slice a view and
+            # therefore updated automatically.
+            # For other types, we update it manually here.
+            dims = self._slice.dims
+            point = np.round(self.world_to_data(dims.point)).astype(int)
+            point_not_displayed = [point[i] for i in dims.not_displayed]
+            indices_not_displayed = [indices[i] for i in dims.not_displayed]
```

Comment:
This new code is quite long. Could I ask to move it to a separate method to increase readability?

Reply:
Good call — I worked on it progressively so it was always "one more line" so I didn't notice 😅. Will push shortly.